File: lib/domain/user/user_repository.dart
import 'package:dartz/dartz.dart';

import 'user_entity.dart';
import 'user_failures.dart';

abstract class UserRepository {
  Future<Either<UserFailure, Option<UserEntity>>> getCurrentUser();
  Future<Either<UserFailure, List<UserEntity>>> searchUsers({String? query});
}

File: backend/src/core/domain/users/user_repository.ts
import { Either, Option } from "funpjs";
import UserEntity from "./user_entity";
import UserFailure from "./user_failure";

export type UserQueryOptions = {
  email?: string;
  token?: string;
  uid?: string;
};

export type UserUpdateOptions = {
  password?: string;
  name?: string;
  pushFriend?: string;
};

export default interface UserRepository {
  getUser(
    query: UserQueryOptions
  ): Promise<Either<UserFailure, Option<UserEntity>>>;
  createUser(
    email: string,
    password: string
  ): Promise<Either<UserFailure, UserEntity>>;
  exists(query: UserQueryOptions): Promise<boolean>;
  getAllUsers(): Promise<UserEntity[]>;
  searchUsers(query: string): Promise<UserEntity[]>;
  updateUser(
    id: string,
    options: UserUpdateOptions
  ): Promise<Either<UserFailure, UserEntity>>;
}
